### PR TITLE
UCP/CORE/WIREUP: Use UCS conn match

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Use UCS conn match in UCP

## Why ?

To avoid duplication of the connection matching functionality in UCP/UCTs

## How ?

1. Remove EP match form UCP/WIREUP code
2. Add adoption of UCS connection matching

### CODE REVIEW

File: src/ucp/core/ucp_worker.c [hunk 1]
```diff
@@ -609,6 +609,113 @@ ucs_status_t ucp_worker_set_ep_failed(ucp_worker_h worker, ucp_ep_h ucp_ep,
     return ret_status;
 }
 
+const void *ucp_worker_ep_match_get_address(const ucs_conn_match_elem_t *conn_match)
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:612]: maybe keep the ep_match.c/h files, and put the wrapper functions there?

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:612]: > maybe keep the ep_match.c/h files, and put the wrapper functions there?

@yosefe yes, it's better

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:615]: @yosefe fixed

File: src/ucp/core/ucp_ep.h [hunk 2]
```diff
@@ -369,6 +369,16 @@ typedef struct {
...
 } ucp_ep_close_proto_req_t;
 
+
+/**
+ * Object that represents matching with remote endpoints
+ */
+typedef struct {
+    uint64_t                  dest_uuid;         /* Destination worker UUID */
+    ucs_conn_match_elem_t     conn_match;        /* Connection matching object */
+} ucp_ep_match_t;
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.h:379]: ucp_ep_match_elem_t

File: src/ucp/core/ucp_worker.c [hunk 3]
```diff
@@ -1782,9 +1782,10 @@ ucs_status_t ucp_worker_create(ucp_context_h context,
     ucs_list_head_init(&worker->arm_ifaces);
     ucs_list_head_init(&worker->stream_ready_eps);
     ucs_list_head_init(&worker->all_eps);
-    ucp_ep_match_init(&worker->ep_match_ctx);
+    ucs_conn_match_init(&worker->conn_match_ctx, sizeof(uint64_t),
+                        &ucp_ep_match_ops);
 
-    UCS_STATIC_ASSERT(sizeof(ucp_ep_ext_gen_t) <= sizeof(ucp_ep_t));
+    //UCS_STATIC_ASSERT(sizeof(ucp_ep_ext_gen_t) <= sizeof(ucp_ep_t));
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:1788]: leftover?

File: src/ucp/wireup/ep_match.c [hunk 4]
```diff
@@ -8,174 +8,113 @@
...
-        if (!ucs_hlist_is_empty(&entry.unexp_ep_q)) {
-            ucs_warn("match_ctx %p: uuid 0x%"PRIx64" unexpected queue is not empty",
-                     match_ctx, dest_uuid);
-        }
-    })
-    kh_destroy_inplace(ucp_ep_match, &match_ctx->hash);
+    return (ucs_conn_sn_t)ucp_ep_from_ext_gen((ucp_ep_ext_gen_t*)
+                                              ucs_container_of(conn_match,
+                                                               ucp_ep_ext_gen_t,
+                                                               ep_match.conn_match))->conn_sn;
```
Comment (yosefe) [inline src/ucp/wireup/ep_match.c:31]: helper function ucp_ep_ext_gen_from_conn_match ?

File: src/ucp/wireup/ep_match.c [hunk 5]
```diff
@@ -8,174 +8,113 @@
...
-        /* initialize match list on first use */
-        entry->next_conn_sn    = 0;
-        ucs_hlist_head_init(&entry->exp_ep_q);
-        ucs_hlist_head_init(&entry->unexp_ep_q);
+    size_t required_size = snprintf(NULL, 0, "%zu", *(uint64_t*)address);
+
+    if (max_size < required_size) {
+        ucs_fatal("size of the string (%zu) is not enough to be filled"
+                  " by the address (%zu), required - %zu",
+                  max_size, *(uint64_t*)address, required_size);
```
Comment (yosefe) [inline src/ucp/wireup/ep_match.c:42]: better print as much as fits in

File: src/ucp/wireup/ep_match.c [hunk 6]
```diff
@@ -8,174 +8,113 @@
...
-    kh_destroy_inplace(ucp_ep_match, &match_ctx->hash);
+    return (ucs_conn_sn_t)ucp_ep_from_ext_gen((ucp_ep_ext_gen_t*)
+                                              ucs_container_of(conn_match,
+                                                               ucp_ep_ext_gen_t,
+                                                               ep_match.conn_match))->conn_sn;
 }
 
-static ucp_ep_match_entry_t*
-ucp_ep_match_entry_get(ucp_ep_match_ctx_t *match_ctx, uint64_t dest_uuid)
+const char* ucp_ep_match_address_str(const void *address,
```
Comment (yosefe) [inline src/ucp/wireup/ep_match.c:34]: static

File: src/ucp/wireup/ep_match.c [hunk 7]
```diff
@@ -8,174 +8,113 @@
...
 {
-    kh_init_inplace(ucp_ep_match, &match_ctx->hash);
+    const ucp_ep_ext_gen_t *ep_ext = ucs_container_of(conn_match,
+                                                      ucp_ep_ext_gen_t,
+                                                      ep_match.conn_match);
+    return &ep_ext->ep_match.dest_uuid;
 }
 
-void ucp_ep_match_cleanup(ucp_ep_match_ctx_t *match_ctx)
+ucs_conn_sn_t ucp_ep_match_get_conn_sn(const ucs_conn_match_elem_t *conn_match)
```
Comment (yosefe) [inline src/ucp/wireup/ep_match.c:26]: static

File: src/ucp/wireup/ep_match.c [hunk 8]
```diff
@@ -8,174 +8,113 @@
...
 #include <ucp/core/ucp_ep.inl>
-#include <ucp/wireup/ep_match.h>
-#include <inttypes.h>
 
 
-__KHASH_IMPL(ucp_ep_match, static UCS_F_MAYBE_UNUSED inline, uint64_t,
-             ucp_ep_match_entry_t, 1, kh_int64_hash_func, kh_int64_hash_equal);
-
-void ucp_ep_match_init(ucp_ep_match_ctx_t *match_ctx)
+const void *ucp_ep_match_get_address(const ucs_conn_match_elem_t *conn_match)
```
Comment (yosefe) [inline src/ucp/wireup/ep_match.c:18]: static

File: src/ucp/wireup/ep_match.c [hunk 9]
```diff
@@ -8,174 +8,113 @@
...
-            ucs_trace("match_ctx %p: matched %s ep %p by uuid 0x%"PRIx64" conn_sn %d",
-                      match_ctx, title, ep, dest_uuid, conn_sn);
-            ucs_assertv(ucs_test_all_flags(ep->flags,
-                                           exp_ep_flags | UCP_EP_FLAG_ON_MATCH_CTX),
-                        "ep=%p flags=0x%x exp_flags=0x%x", ep, ep->flags,
-                        exp_ep_flags);
-            ep->flags &= ~UCP_EP_FLAG_ON_MATCH_CTX;
-            return ep;
+        ucp_ep_flags_t UCS_V_UNUSED exp_ep_flags = UCP_EP_FLAG_ON_MATCH_CTX |
+                                                   (!is_exp ? UCP_EP_FLAG_DEST_EP : 0);
```
Comment (yosefe) [inline src/ucp/wireup/ep_match.c:89]: (is_exp ? 0 : UCP_EP_FLAG_DEST_EP);

File: src/ucp/core/ucp_ep.h [hunk 10]
```diff
@@ -369,6 +369,16 @@ typedef struct {
                                                    used in close protocol */
 } ucp_ep_close_proto_req_t;
 
+
+/**
+ * Object that represents matching with remote endpoints
+ */
+typedef struct {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.h:376]: move it to ep_match.h

Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h:376]: done

File: src/ucp/core/ucp_ep.h (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h:379]: done

File: src/ucp/core/ucp_worker.c (hunk 3 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:1788]: yes, removed

File: src/ucp/wireup/ep_match.c (hunk 8 above)
Comment (dmitrygx) [reply src/ucp/wireup/ep_match.c:18]: done

File: src/ucp/wireup/ep_match.c (hunk 7 above)
Comment (dmitrygx) [reply src/ucp/wireup/ep_match.c:26]: done

File: src/ucp/wireup/ep_match.c (hunk 6 above)
Comment (dmitrygx) [reply src/ucp/wireup/ep_match.c:34]: done

File: src/ucp/wireup/ep_match.c (hunk 4 above)
Comment (dmitrygx) [reply src/ucp/wireup/ep_match.c:31]: done

File: src/ucp/wireup/ep_match.c (hunk 9 above)
Comment (dmitrygx) [reply src/ucp/wireup/ep_match.c:89]: done

**Pull request merged**